src/core/queryClient.ts

```diff
@@ -532,13 +533,35 @@ export class QueryClient {
     }
   }
 
+  findFirstMatchingQueryDefaults(
+    queryKey?: QueryKey
+  ): QueryOptions<any, any, any> | undefined {
+    if (!queryKey) {
+      return undefined
+    }
+
+    // First retrieve all matching defaults for the given key
+    const matchingDefaults = this.queryDefaults.filter(x =>
+      partialMatchKey(queryKey, x.queryKey)
+    )
+    // It is ok not having defaults, but it is error prone to have more than 1 default for a given key
+    if (matchingDefaults?.length > 1) {
+      getLogger().warn(
+        `[QueryClient] Several defaults match with key '${JSON.stringify(
+          queryKey
+        )}'. The first matching query defaults are used. Please check how query defaults are registered. Order does matter here. cf. https://react-query.tanstack.com/reference/QueryClient#queryclientsetquerydefaults.`
+      )
+    }
```

Comment:
let's please hide this behind a `if (process.env.NODE_ENV !== 'production')` check

Reply:
Yep, I removed it but can't remember why…

Reply:
New `assert` helper in utils to hide all this.